cpp/src/lists/concatenate_rows.cu

```diff
@@ -0,0 +1,446 @@
+                          }
+                        }
+                        d_validities[idx] = static_cast<int8_t>(is_valid);
+                      }
+                      return d_interleaved_offsets[idx * num_cols];
+                    });
+
+  // The child column containing list entries is taken from the `interleaved_columns` column.
+  auto interleaved_columns_content = interleaved_columns->release();
+
+  rmm::device_buffer null_mask{};
+  size_type null_count{0};
+  if (has_null_mask) {
+    std::tie(null_mask, null_count) = cudf::detail::valid_if(
+      validities.begin(), validities.end(), thrust::identity<int8_t>{}, stream, mr);
+    if (null_count == 0) {
+      // The output from `valid_if` is always a full-size buffer.
+      null_mask = rmm::device_buffer{};
+    }
+  }
+
+  return make_lists_column(
+    num_output_lists,
+    std::move(list_offsets),
+    std::move(interleaved_columns_content.children[lists_column_view::child_column_index]),
+    null_count,
+    std::move(null_mask),
+    stream,
+    mr);
+}
+
+/**
+ * @brief Generate list offsets and list validities for the output lists column from the table_view
+ * of the input lists columns.
+ *
+ * This function is called only when has_null_mask == true and
+ * null_policy == concatenate_null_policy::NULLIFY_OUTPUT_ROW.
+ */
+std::pair<std::unique_ptr<column>, rmm::device_uvector<int8_t>>
+generate_list_offsets_and_validities(table_view const& input,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr)
+{
+  auto const num_cols         = input.num_columns();
+  auto const num_output_lists = input.num_rows();
+  auto const table_dv_ptr     = table_device_view::create(input);
+
+  // The output offsets column.
+  static_assert(sizeof(offset_type) == sizeof(int32_t));
+  static_assert(sizeof(size_type) == sizeof(int32_t));
+  auto list_offsets = make_numeric_column(
+    data_type{type_id::INT32}, num_output_lists + 1, mask_state::UNALLOCATED, stream, mr);
+  auto const d_offsets = list_offsets->mutable_view().begin<offset_type>();
+
+  // The array of int8_t to store validities for list elements.
+  auto validities = rmm::device_uvector<int8_t>(num_output_lists, stream);
+
+  // Compute list sizes and validities.
+  thrust::transform(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator<size_type>(0),
+    thrust::make_counting_iterator<size_type>(num_output_lists),
+    d_offsets,
+    [num_cols, table_dv = *table_dv_ptr, d_validities = validities.begin()] __device__(
+      size_type const idx) {
+      size_type list_size{0};
+      bool is_valid{true};
+      for (size_type col_id = 0; col_id < num_cols; ++col_id) {
+        auto const& lists_col = table_dv.column(col_id);
+        if (lists_col.is_null(idx)) {
+          is_valid  = false;
+          list_size = 0;
+          break;
+        }
+        auto const list_offsets =
```

Comment:
If you do replace the loop with a `thrust::all_of` then this can become a `thrust::reduce`